2020 Jamstack Chapter (#1632)
HTTPArchive/almanac.httparchive.org: src/content/en/2020/jamstack.md

part_number: III
chapter_number: 17
title: Jamstack
description: Jamstack chapter of the 2020 Web Almanac covering the use of Jamstack, the performance of popular Jamstack frameworks, as well as an analysis of real-user experience using the Core Web Vitals metadata.
authors: [ahmadawais]

> analysts: [denar90, remotesynth]
> reviewers: [maedahbatool, phacks]

Reviewer: Order is reviewers and then analsyts:

```suggestion
reviewers: [MaedahBatool, phacks]
analysts: [denar90, remotesynth]
```

@MaedahBatool / @remotesynth we didn;t have a profile for you so added the following in https://github.com/HTTPArchive/almanac.httparchive.org/pull/1632/commits/95c230bd8579e512967111d55e3a1e6a22971453:

```

    "remotesynth": {
      "name": "Brian Rinaldi",
      "teams": [
        "analysts"
      ],
      "avatar_url": "https://avatars1.githubusercontent.com/u/216712?v=4&s=200",
      "website": "https://remotesynthesis.com/",
      "github": "remotesynth",
      "twitter": "remotesynth"
    },
```

```
    "MaedahBatool": {
      "name": "Maedah Batool",
      "teams": [
        "reviewers"
      ],
      "avatar_url": "https://avatars1.githubusercontent.com/u/12712988?v=4&s=200",
      "website": "https://maedahbatool.com/",
      "github": "MaedahBatool",
      "twitter": "maedahbatool"
    },
```

Let us know if you want any of that changed?

discuss: 2053
results: https://docs.google.com/spreadsheets/d/1BCC5Q4tePpTl8TiaGmSxBc9Lh2to7xBfVPMULFOBwvk/
queries: 17_Jamstack

> featured_quote: Stats suggest almost twice as many Jamstack sites exist now than in 2019. Developers enjoy a better development experience but that doesn't always translate to a better user experience. Let’s work towards continued progress for improving the real-user experience of browsing Jamstack sites.

Reviewer: This is a bit of a mismash of many quotes. What about this:

```suggestion
featured_quote: Stats suggest almost twice as many Jamstack sites exist now than in 2019. Developers enjoy a better development experience by separating the frontend from the backend. But what about the real-user experience of browsing Jamstack sites?
```

> featured_stat_1: 154%
> featured_stat_label_1: more mobile websites built with Jamstack in 2020
> featured_stat_2: 1 gram
> featured_stat_label_2: CO2 is emitted by the median Jamstack page load
> featured_stat_3: 38%
> featured_stat_label_3: Jamstack sites are built using Next.js (three times more than any other framework!)

Reviewer: This is what it looks like:

![Featured Stats](https://user-images.githubusercontent.com/10931297/101237541-8ad1c480-36d1-11eb-8250-b553d8c50c87.png)

Change to label sense, and shorten:

```suggestion
featured_stat_1: 154%
featured_stat_label_1: Increase in Jamstack websites in 2020
featured_stat_2: 1 gram
featured_stat_label_2: CO2 emitted the median Jamstack on page load
featured_stat_3: 38%
featured_stat_label_3: Jamstack sites built using Next.js
```

This would look like this:

![Feratured Stats (editted)](https://user-images.githubusercontent.com/10931297/101237643-6fb38480-36d2-11eb-9bfc-a6d11c893ca4.png)

---

## Introduction

Jamstack is a relatively new concept of an architecture designed to make the web faster, more secure, and easier to scale. It builds on many of the tools & workflows which developers love and which bring maximum productivity.

The core principles of Jamstack are pre-rendering your site pages and decoupling the frontend from the backend. It relies on the idea of delivering the frontend content hosted separately on a CDN provider that uses APIs (for example, a headless CMS) as its backend if any.

> The [HTTP Archive][1] crawls [millions of pages][2] every month and runs them through a private instance of [WebPageTest][3] to store key information on every page. (You can learn more about this in our [methodology][4]). In the context of Jamstack, HTTP Archive provides extensive information on the usage of the frameworks and CDNs for the entire web. This chapter consolidates and analyzes many of these trends.

Reviewer: We just inline the URLs in place rather than at the end:

```suggestion
The [HTTP Archive](https://httparchive.org/) crawls [millions of pages](https://httparchive.org/reports/state-of-the-web#numUrls) every month and runs them through a private instance of [WebPageTest](https://webpagetest.org/) to store key information on every page. (You can learn more about this in our [methodology](./methodology)). In the context of Jamstack, HTTP Archive provides extensive information on the usage of the frameworks and CDNs for the entire web. This chapter consolidates and analyzes many of these trends.
```

> ### Goals: Growth of Jamstack & Real User Perf

Reviewer: Quite long and repeating what's in the next paragraph. To be honest I think it can be removed:

```suggestion
```

(Getting into editing here, but not much time left before launch so going to make some light editing suggestions even at this stage)

The goals of this chapter are to estimate and analyze the growth of the Jamstack sites, the performance of popular Jamstack frameworks, as well as an analysis of real user experience using the Core Web Vitals metadata.

> ## Adoption: What is the overall level of Jamstack adoption amongst all websites, across desktop and mobile?

Reviewer: This is VERY long. Look at what it looks like in the table of contents:

![Jamstack ToC](https://user-images.githubusercontent.com/10931297/101237719-f4060780-36d2-11eb-80d5-865adb5cd0ac.png)

```suggestion
## Adoption of Jamstack adoption
```

Our analysis throughout this work looks at desktop and mobile websites. The vast majority of URLs we looked at are in both datasets, but some URLs are only accessed by desktop or mobile devices. This can cause small divergences in the data, and we thus look at desktop and mobile results separately.

{{ figure_markup(
  image="jamstack-adoption.png",
  caption="Jamstack adoption trend.",
  description="Bar chart showing the increase in the level of Jamstack adoption in 2019 and 2020. Mobile has increased from 0.40% to 0.72%. Desktop has increased from 0.27% to 0.68%.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQ-zBygubF5NuAXOGScVLhnozB_6Pi_0z4B57cP_I6AfeBuZCGzs3aP_1nPjPp37aNRATRYC6EfOeYX/pubchart?oid=1650360073&format=interactive",
  sheets_gid="908645975",
  sql_file="ssg_compared_to_2019.sql"
  )

> }}

Reviewer: We need to drop the 3D effect - it's inconsistent with the other chapters. I can take care of this and redoing the images so just an FYI for you.

Reviewer: Charts updated in https://github.com/HTTPArchive/almanac.httparchive.org/pull/1632/commits/7f19a62cbac53429817b3b1f649d377633103af2 to:
- Remove 3D effect
- Switch to sentence case
- Use standard height and widths (600 x 371)
- Move CWV labels to outside bar chart as needed when using standard height and widths
- Export hi-res images

More than 0.7% of web pages are powered by Jamstack, an increase of over 79% from 2019. This breaks down to 0.72% on desktop, up from 0.40%  in 2019, and 0.68% on mobile, up from 0.27% in 2019. We believe this is a sign of the steady growth of the Jamstack community. You can read more about ours on the [Methodology][4] page.

<figure markdown>

|   Year   | Desktop | Mobile |
| -------- | ------- | ------ |
| 2019     | 0.40%   | 0.27%  |
| 2020     | 0.72%   | 0.68%  |
| % Change | 79%     | 154%   |

<figcaption>{{ figure_link(caption="Jamstack adoption statistics.", sheets_gid="908645975", sql_file="ssg_compared_to_2019.sql") }}</figcaption>
</figure>

The increase in desktop web pages powered by a Jamstack framework is 79% from last year. On mobile, this increase is almost two folds, at 154%. This is a significant growth from 2019, especially for mobile pages.

> ## Top SSGs YoY: Which are the most common Jamstack frameworks across desktop and mobile sites?

Reviewer: Again, very long. And avoid use of Acronyms before you explain what they mean:

```suggestion
## Top Jamstacks?
```

> Our analysis counted 12 separate Jamstack frameworks. Only five frameworks had more than 1% share. Next.js, Gatsby, Hugo, Jekyll being the top contenders for the Jamstack market share.

Reviewer: I think we need to talk about "the Eleventy problem" here. Or in the introduction.

```suggestion
Our analysis counted 12 separate Jamstack frameworks. Only five frameworks had more than 1% share. Next.js, Gatsby, Hugo, Jekyll being the top contenders for the Jamstack market share.

<p class="note">It should be noted that our analysis is limited by those Jamstacks that make themselves easily identified using [Wappalyzer](./methodology#wappalyzer). This means our data does not include some popular Jamstacks like [Eleventy](https://github.com/11ty/eleventy/) which [make a deliberate choice to not make themselves identifiable](https://twitter.com/eleven_ty/status/1334225624110608387?s=20). While we would ideally include all Jamstacks, we believe there is still value in analysing what we have</p>
```

  sheets_gid="364341150",
  sql_file="ssg_compared_to_2019.sql"
  )
}}

It's worth noting here the fact that Next.js websites include a mix of both Statically Generated Sites (SSG) pages and Server Side Rendered (SSR) pages. This is due to the lack of our ability to measure them separately. This means that the analysis may include sites that are mostly or partially server-rendered, meaning they do not fall under the traditional definition of a Jamstack site. Nonetheless, it appears that this hybrid nature of Next.js gives it a competitive advantage over other frameworks hence making it more popular.

{{ figure_markup(
  image="jamstack-adoption-share-2020-pie.png",
  caption="Jamstack adoption share pi chart 2020.",
  description="Pie chart showing the adoption of Jamstack frameworks dominated by Next.js taking three times the share of 38% then the next most popular one, Gatsby with ~12% share.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQ-zBygubF5NuAXOGScVLhnozB_6Pi_0z4B57cP_I6AfeBuZCGzs3aP_1nPjPp37aNRATRYC6EfOeYX/pubchart?oid=115327297&format=interactive",
  sheets_gid="364341150",
  sql_file="ssg_compared_to_2019.sql"
  )
}}

> ## What’s the impact of Jamstack on the environment?

Reviewer: ```suggestion
## Environmental impact
```

> This year we have sought to better understand the impact of Jamstack sites on the environment. [The information and communications technology (ICT) industry accounts for 2% of global carbon emissions][6], and data centers specifically account for 0.3% of global carbon emissions. This puts the ICT industry’s carbon footprint equivalent to the aviation industry’s emissions from fuel.

Reviewer: ```suggestion
This year we have sought to better understand the impact of Jamstack sites on the environment. [The information and communications technology (ICT) industry accounts for 2% of global carbon emissions](https://www.nature.com/articles/d41586-018-06610-y), and data centers specifically account for 0.3% of global carbon emissions. This puts the ICT industry’s carbon footprint equivalent to the aviation industry’s emissions from fuel.
```

Jamstack is often credited for being mindful of performance. In the next section, we look into the carbon emissions of Jamstack websites.

### Page Weight

> Our research looked at the average Jamstack page weight in KB and mapped this to CO2 emissions using logic from the [Carbon API][7]. This generated the following results, split by desktop and mobile:

Reviewer: ```suggestion
Our research looked at the average Jamstack page weight in KB and mapped this to CO2 emissions using logic from the [Carbon API](https://gitlab.com/wholegrain/carbon-api-2-0/-/blob/master/includes/carbonapi.php#L342). This generated the following results, split by desktop and mobile:
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQ-zBygubF5NuAXOGScVLhnozB_6Pi_0z4B57cP_I6AfeBuZCGzs3aP_1nPjPp37aNRATRYC6EfOeYX/pubchart?oid=1748236124&format=interactive",
  sheets_gid="881269360",
  sql_file="distribution_of_page_weight_requests_and_co2_grams_per_ssg_web_page.sql"
  )
}}

We found that the median Jamstack page load resulted in the transfer of 1.5 MB of various assets and thus the emission of 1.1 grams of CO2 for desktop and 0.9 grams for mobile. This was the same for desktop and mobile. The most efficient percentile of Jamstack web pages result in the generation of at least one third less CO2 than the median, whilst the least efficient percentile of Jamstack web pages goes the other way, generating around four times more.

Page weights are important here. The average desktop Jamstack web page loads 1.5 MB of video, image, script, font, CSS, and audio data. 10% of pages, however, load over 4 MB of this data. On mobile devices, the average web page loads 0.7 MB fewer than on desktop, a fact consistent across all percentiles.

### Image formats

> Popular image formats are PNG, JPG, GIF, SVG, WebP, and ICO. Of these, [WebP is the most efficient in most situations][8], with WebP lossless images [26% smaller][9] than equivalent PNGs and [25-34% smaller][10] than comparable JPGs. We see, however, that WebP is the second least popular image format across all Jamstack pages, where PNG is the most popular both for mobile and desktop. Only slightly less popular is JPG whereas GIF is almost 20% of all the images used on Jamstack sites. An interesting discovery is SVG which is almost twice as popular on mobile sites as desktop sites.

Reviewer: ```suggestion
Popular image formats are PNG, JPG, GIF, SVG, WebP, and ICO. Of these, [WebP is the most efficient in most situations](https://developers.google.com/speed/webp/), with WebP lossless images [26% smaller](https://developers.google.com/speed/webp/docs/webp_lossless_alpha_study#results) than equivalent PNGs and [25-34% smaller](https://developers.google.com/speed/webp/docs/webp_study) than comparable JPGs. We see, however, that WebP is the second least popular image format across all Jamstack pages, where PNG is the most popular both for mobile and desktop. Only slightly less popular is JPG whereas GIF is almost 20% of all the images used on Jamstack sites. An interesting discovery is SVG which is almost twice as popular on mobile sites as desktop sites.
```

{{ figure_markup(
  image="jamstack-third-party-bytes.png",
  caption="Third party bytes.",
  description="Bar chart showing the amount of third party bytes (in KB) for each percentile for Jamstack pages. At the 10th percentile it is 44.64 KB for desktop and 66.69 for mobile, at the 25th percentile it is 150.63 KB for mobile and 245.78 KB for mobile, at the 50% it's 506.27 KB for desktop and 823.56 KB for mobile, at the 75th percentile 1,340.39 KB for desktop and 1,989.24 for mobile and at the 90th percentile it is 3,169.14 KB for desktop and 3,290.07 for mobile.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQ-zBygubF5NuAXOGScVLhnozB_6Pi_0z4B57cP_I6AfeBuZCGzs3aP_1nPjPp37aNRATRYC6EfOeYX/pubchart?oid=1779247936&format=interactive",
  sheets_gid="1292933800",
  sql_file="third_party_bytes_and_requests_on_ssgs.sql"
  )
}}

We find that the median desktop Jamstack page has 28 third party requests with 506 KB of content, with the mobile equivalent generating 46 requests with 824 KB of content. Whereas 10% of the desktop sites have 132 requests with 3.1MB of content which is only superseded by 168 requests on mobile with 3.2MB of content.

> ## Core Web Vitals: what is the user experience on Jamstack websites?

Reviewer: ```suggestion
## User experience
```

> Jamstack websites are often said to offer a good user experience. It's what the entire concept of separating the frontend from the backend and hosting it on the CDN edge is all about. We aim to shed light on real-world user experience when using Jamstack websites. It's what we call, the [Core Web Vitals][11].

Reviewer: ```suggestion
Jamstack websites are often said to offer a good user experience. It's what the entire concept of separating the frontend from the backend and hosting it on the CDN edge is all about. We aim to shed light on real-world user experience when using Jamstack websites. It's what we call, the [Core Web Vitals](https://web.dev/learn-web-vitals/).
```

We’ve reviewed the adoption of Jamstack, user experience of websites created by these Jamstack frameworks, and for the first time looked at the impact of Jamstack on the environment. We have answered many questions here but leave further questions unanswered.

There are frameworks like Eleventy which we weren't able to measure or analyze since there is no pattern available to determine the usage of such frameworks, which has an impact on the data presented here. Next.js offers both Static Generation and Server Rendering, separating the two in this data is nearly impossible since it also offers incremental Static Generation. Further research building on this chapter will be gratefully received.

Moreover, we have highlighted some areas which need attention from the Jamstack community. We hope there will be progress to share in the 2021 report. Different Jamstack frameworks can start to document how to improve real user experience by looking at Core Web Vitals.

> Vercel, one of the CDNs meant to host Jamstack sites, has built an analytics offering called [Real User Experience Score][12]. While other performance measuring tools like [Lighthouse][13] estimate your user's experience by running a simulation in a lab, Vercel's Real Experience Score is calculated using real data points collected from the devices of the actual users of your application.

Reviewer: ```suggestion
Vercel, one of the CDNs meant to host Jamstack sites, has built an analytics offering called [Real User Experience Score](https://vercel.com/docs/analytics#real-experience-score). While other performance measuring tools like [Lighthouse](https://web.dev/measure/) estimate your user's experience by running a simulation in a lab, Vercel's Real Experience Score is calculated using real data points collected from the devices of the actual users of your application.
```

Probably worth noting here that Vercel created and maintains Next.js, particularly since Next.js had low LCP scores. This could mean we can hope to see a marked improvement next year. This can be extremely helpful information for developers.

Jamstack frameworks are improving the developer experience of building sites. Let’s work towards continued progress for improving the real-user experience of browsing Jamstack sites.

> 
> [1]: https://httparchive.org/
> [2]: https://httparchive.org/reports/state-of-the-web#numUrls
> [3]: https://webpagetest.org/
> [4]: ./methodology
> [6]: https://www.nature.com/articles/d41586-018-06610-y
> [7]: https://gitlab.com/wholegrain/carbon-api-2-0/-/blob/master/includes/carbonapi.php#L342
> [8]: https://developers.google.com/speed/webp/
> [9]: https://developers.google.com/speed/webp/docs/webp_lossless_alpha_study#results
> [10]: https://developers.google.com/speed/webp/docs/webp_study
> [11]: https://web.dev/learn-web-vitals/
> [12]: https://vercel.com/docs/analytics#real-experience-score
> [13]: https://web.dev/measure/

Reviewer: ```suggestion
```
